style: :dir() pseudo class now represented by enum

wilsoniya opened the pull request:
`:dir()` pseudo class param now represented as enum variants.

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #16840
- [X] There are tests for these changes

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/19195)
<!-- Reviewable:end -->

wilsoniya committed:
style: :dir() pseudo class now represented by enum

Fixes #19195

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @mbrubeck (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/selectors/size_of_tests.rs, components/style/invalidation/element/invalidation_map.rs, components/style/gecko/selector_parser.rs, components/style/invalidation/element/element_wrapper.rs, components/style/gecko/wrapper.rs and 2 more
 * @canaltinova: components/style/invalidation/element/invalidation_map.rs, components/style/gecko/selector_parser.rs, components/style/invalidation/element/element_wrapper.rs, components/style/gecko/wrapper.rs, components/style/gecko/non_ts_pseudo_class_list.rs
 * @emilio: components/style/invalidation/element/invalidation_map.rs, components/style/gecko/selector_parser.rs, components/style/invalidation/element/element_wrapper.rs, components/style/gecko/wrapper.rs, components/style/gecko/non_ts_pseudo_class_list.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

emilio commented on components/selectors/gecko_like_types.rs:
```diff
@@ -6,10 +6,18 @@
+pub enum Direction {
+    Rtl,
+    Ltr,
+    Other(String),
```
This should still be `Box<[u16]>`, or `Box<str>` at least. That may avoid `Component` growing.

emilio commented on components/selectors/size_of_tests.rs:
```diff
@@ -14,8 +14,8 @@ use visitor::SelectorVisitor;
 
-size_of_test!(size_of_component, Component<Impl>, 32);
-size_of_test!(size_of_pseudo_class, PseudoClass, 24);
+size_of_test!(size_of_component, Component<Impl>, 48);
```
Growing `Component` is not really acceptable I think :(

emilio commented on components/style/gecko/selector_parser.rs:
```diff
@@ -34,6 +34,52 @@ bitflags! {
+    /// left-to-right semantic directionality
+    Ltr,
+    /// Some other provided directionality value
+    Other(String),
```
Ditto.

emilio commented on components/style/gecko/selector_parser.rs:
```diff
@@ -34,6 +34,52 @@ bitflags! {
+    Other(String),
+}
+
+impl <'a> From<&'a str> for Direction {
```
No need for `From<&'a str>`, let's use the `Parse` trait.

emilio commented on components/style/gecko/selector_parser.rs:
```diff
@@ -34,6 +34,52 @@ bitflags! {
+
+impl <'a> From<&'a str> for Direction {
+    fn from(string: &'a str) -> Direction {
+        match string.to_lowercase().as_str() {
```
No need to allocate here, let's use `match_ignore_ascii_case`.

emilio commented on components/style/gecko/selector_parser.rs:
```diff
@@ -360,6 +417,10 @@ impl<'a, 'i> ::selectors::Parser<'i> for SelectorParser<'a> {
                         NonTSPseudoClass::$k_name(utf16.into_boxed_slice())
                     }, )*
+                    "dir" => {
+                        let name: &str = parser.expect_ident()?;
```
Let's add the `match` here instead of the `From` impl.

emilio commented on components/style/gecko/wrapper.rs:
```diff
@@ -2066,6 +2065,15 @@ impl<'le> ::selectors::Element for GeckoElement<'le> {
             }
+            NonTSPseudoClass::Dir(ref dir) => {
+                unsafe {
+                    Gecko_MatchStringArgPseudo(
```
The point of converting to an `enum` is not having to go through `Gecko_MatchStringArgPseudo`.

Instead of that, we can just do:

```rust
match *dir {
   Direction::Ltr => self.get_state().intersects(element_state::IN_HAS_DIR_ATTR_LTR_STATE),
   Direction::Rtl => self.get_state().intersects(element_state::IN_HAS_DIR_ATTR_RTL_STATE),
   Direction::Other(..) => false,
}
```

emilio commented on components/style/invalidation/element/element_wrapper.rs:
```diff
@@ -182,9 +182,9 @@ impl<'a, E> Element for ElementWrapper<'a, E>
-            NonTSPseudoClass::Dir(ref s) => {
+            NonTSPseudoClass::Dir(ref dir) => {
                 use invalidation::element::invalidation_map::dir_selector_to_state;
-                let selector_flag = dir_selector_to_state(s);
```
Similarly, we can simplify `dir_selector_to_state` to take `&Direction`, instead of comparing the string.

emilio replied:
That should make the `Box<[u16]>` impl unnecessary.

emilio commented on tests/unit/stylo/dir_repr.rs:
```diff
@@ -0,0 +1,31 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
No need for this, we have tests in mozilla-central for this :)

emilio reviewed (changes requested):
Good start!

I have a bunch of comments, let me know if any of those is unclear and I'm happy to clarify or help out.

Thanks for working on this!

wilsoniya commented on components/selectors/size_of_tests.rs:
```diff
@@ -14,8 +14,8 @@ use visitor::SelectorVisitor;
 
-size_of_test!(size_of_component, Component<Impl>, 32);
-size_of_test!(size_of_pseudo_class, PseudoClass, 24);
+size_of_test!(size_of_component, Component<Impl>, 40);
```
Even using the `Other<Box[u16]>` enum variant type, the size of these structs is increased :(

emilio replied:
Hmm, that's pretty unfortunate, let's box the `Direction` then:

```rust
Dir(Box<Direction>)
```

Mentioning that it allows not growing `Component`, even though it's slightly unfortunate.

emilio replied:
Oh, the other thing we can do instead, and it'll fix the serialization bug I just noted below too, I think, is using `CustomIdent`, which should just be one word and allow us not to box it:

```rust
enum Direction {
    Rtl,
    Ltr,
    Other(CustomIdent),
}
```

Then you don't even need to manually write the serialization code, and you can just `#[derive(ToCss)]` on it.

wilsoniya commented:
Hey, @emilio, thanks for the review! I think I've addressed your comments, with the exception of the change in size of `Component`.

Do you have ideas regarding how to eliminate the size increase on that struct?

emilio commented on components/style/selector_parser.rs:
```diff
@@ -173,3 +173,26 @@ impl<T> PerPseudoElementMap<T> {
+    Other(Box<[u16]>),
+}
+
+impl<'a> From<&'a Direction> for String {
```
Given we don't use the utf16 impl anyway, this could be `Box<str>`, and there'd be no need to allocate a `String` for serialization.

emilio commented on components/style/gecko/selector_parser.rs:
```diff
@@ -92,6 +94,12 @@ impl ToCss for NonTSPseudoClass {
                         dest.write_str(&value)?;
                         return dest.write_char(')')
                     }, )*
+                    NonTSPseudoClass::Dir(ref dir) => {
```
Let's `impl ToCss for Direction`, and use that here, doing just:

```rust
dest.write_str(":dir(")?;
dir.to_css(dest)?;
dest.write_char(')')
```

emilio commented on components/style/gecko/selector_parser.rs:
```diff
@@ -92,6 +94,12 @@ impl ToCss for NonTSPseudoClass {
+                    NonTSPseudoClass::Dir(ref dir) => {
+                        let value: String = dir.into();
+                        dest.write_str(concat!(":dir("))?;
+                        dest.write_str(&value)?;
```
Also, I believe this is not correct, and should do escaping for the identifier... Though it's an existing bug.

wilsoniya commented:
@emilio Thanks for the follow-up. Regarding your [comment regarding escaping](https://github.com/servo/servo/pull/19195#discussion-diff-150824734R100): do you have a link to an issue about existing escaping shortcomings and/or thoughts on how to proceed?

Your attention and patience on this PR is much appreciated! At this point it probably would have been faster for you to implement a fix yourself than to explain all these things to me :)

emilio commented on components/style/gecko/selector_parser.rs:
```diff
@@ -92,6 +94,11 @@ impl ToCss for NonTSPseudoClass {
                         return dest.write_char(')')
                     }, )*
+                    NonTSPseudoClass::Dir(ref dir) => {
+                        dest.write_str(concat!(":dir("))?;
```
Why is there a need for `concat!` here?

emilio commented on components/style/selector_parser.rs:
```diff
@@ -173,3 +173,25 @@ impl<T> PerPseudoElementMap<T> {
+        let dir_str = match *self {
+            Direction::Rtl => "rtl",
+            Direction::Ltr => "ltr",
+            Direction::Other(ref other) => other,
```
Please add a `FIXME` here saying:

> FIXME: This should be escaped as an identifier.

That can be a followup if you want, since that's a bugfix, requires tests and all that.

The short version is that it should use `serialize_identifier`, in order to properly handle escapes. I believe we have tests that test this where you could grab an example... But looking at what `serialize_identifier` handles, something like this:

```css
:dir(\-) { .. }
```

Should be a valid rule, but serializes back to:

```css
:dir(-) { .. }
```

Which isn't an identifier token IIRC.

wilsoniya replied:
Created #19231 which I can pick up as a followup.

emilio reviewed (approved):
Looks great!

r=me with these two nits and the commits squashed. Thanks a lot for working on this!

wilsoniya commented:
@emilio are there any other changes you'd like to see before this is mergeable?

emilio commented on components/style/gecko/selector_parser.rs:
```diff
@@ -92,6 +94,12 @@ impl ToCss for NonTSPseudoClass {
                     }, )*
+                    NonTSPseudoClass::Dir(ref dir) => {
+                        dest.write_str(":dir(")?;
+                        // FIXME: This should be escaped as an identifier; see #19231
```
This comment should probably go in the `to_css` impl, but that can be done in a followup PR.

emilio commented:
Nope, sorry! I had missed the latest update. Feel free to ping me sooner if this happens again.


@bors-servo r+

bors-servo commented:
:pushpin: Commit 10f3ef4 has been approved by `emilio`

<!-- @bors-servo r=emilio 10f3ef42bba3e0123badf3c06426488128ae9f14 -->

emilio commented:
Looks great, thanks a lot!

bors-servo commented:
:hourglass: Testing commit 10f3ef42bba3e0123badf3c06426488128ae9f14 with merge 006202732f0bd8d2239bdd51898380bdbe0f0f1a...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/9856), [arm32](http://build.servo.org/builders/arm32/builds/10090), [arm64](http://build.servo.org/builders/arm64/builds/10120), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10138), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/6969), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/6995), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10197), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/2630), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/2646), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/6891), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/6956), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/2710), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/2670), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/5513)
Approved by: emilio
Pushing 006202732f0bd8d2239bdd51898380bdbe0f0f1a to master...
